File: Python/0238-product-array-except-self.py
#Time: O(n)
#Space: O(1)

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        if not nums:
            return []

        leftProduct = [1 for _ in range(len(nums))]

        for i in range(1,len(nums)):
            leftProduct[i] = leftProduct[i-1]*nums[i-1]

        rightProduct = 1

        for i in reversed(range(len(nums)-1)):
            rightProduct*=nums[i+1]
            leftProduct[i]*=rightProduct

        return leftProduct

File: C++/0238-product-array-except-self.cpp
// Tags: Amazon Array
// Time O(n)
// Space O(1)
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
      if(nums.empty())
      return {};

      vector<int> leftProduct(nums.size());

      leftProduct[0] = 1;

      for(int i=1; i<nums.size(); ++i)
      leftProduct[i] = leftProduct[i-1]*nums[i-1];

      int rightProduct = 1;

      for(int i = nums.size()-2; i>=0 ; --i)
      {
        rightProduct*=nums[i+1];
        leftProduct[i]*=rightProduct;
      }

      return leftProduct;

    }
};
